Added a dockerfile to simplify building and generating snmp configs

tantona opened the pull request:
This PR adds a dockerfile which simplifies building the generator, reading MIB files (mounted in the running container), and generating the resulting snmp.yml.

SuperQ commented on generator/Dockerfile:
```diff
@@ -0,0 +1,20 @@
+FROM golang:1.9.2
```
I would recommend `golang:latest` to pull in the latest stable, rather than a specific release.

SuperQ commented on generator/Dockerfile:
```diff
@@ -0,0 +1,20 @@
+FROM golang:1.9.2
+
+# snmp-mibs-downloader is in the multiverse
```
Nit: s/multiverse/non-free/ (multiverse is an Ubuntu name)

brian-brazil commented on generator/Dockerfile:
```diff
@@ -0,0 +1,20 @@
+RUN apt-get update && apt-get install -y \
+  build-essential \
+  libsnmp-dev \
+  snmp-mibs-downloader
```
You're not actually using this.

brian-brazil commented on generator/Dockerfile:
```diff
@@ -0,0 +1,20 @@
+WORKDIR "/opt"
+
+# Run in a shell - redirect stdout to /dev/null and spit out the resulting config
+CMD /go/bin/generator generate > /dev/null 2>&1 && cat /opt/snmp.yml
```
It'd be useful if other commands could also be run.

tantona commented:
@brian-brazil @SuperQ Thank you for your reviews! 

I've updated the PR so that 
- Removed unused deps from the container
- You can pass whatever command you'd like to the generator
- You can specify the output dir for the snmp config file (I did this so you can mount a volume and write the resulting config file without having to do some wacky redirect)
- Updated the docker usage in the readme with explanations for the volume mounts

tantona commented on generator/Dockerfile:
```diff
@@ -0,0 +1,22 @@
+RUN apt-get update && apt-get install -y libsnmp-dev
+
+# Pull published snmp_exporter dependency so that we have access to
+# shared code in the snmp_exporter project
```
@ build time

brian-brazil commented on generator/Dockerfile:
```diff
@@ -1,20 +1,22 @@
 
+WORKDIR /go/src/github.com/prometheus/snmp_exporter/generator/
+
+# Copy local generator directory so we can develop locally and test
```
I can see this causing problems for normal use, if someone wants to do development they can build the generator by hand.

brian-brazil commented on generator/README.md:
```diff
@@ -24,6 +24,27 @@ The generator reads in from `generator.yml` and writes to `snmp.yml`.
+# Build the image
+docker build -t snmp-generator .
+
+# The following command will:
```
This detail is not relevant to the user. They just need to know where to put the mibs, and where the output will be.

brian-brazil commented on generator/main.go:
```diff
@@ -46,21 +47,22 @@ func generateConfig(nodes *Node, nameToNode map[string]*Node) {
 	generateCommand    = kingpin.Command("generate", "Generate snmp.yml from generator.yml")
 	parseErrorsCommand = kingpin.Command("parse_errors", "Debug: Print the parse errors output by NetSNMP")
 	dumpCommand        = kingpin.Command("dump", "Debug: Dump the parsed and prepared MIBs")
+	outputDir          = kingpin.Flag("output-dir", "Directory to write resulting config file").Default("/opt").Short('o').String()
```
This is a breaking change, it should continue to write to the current directory.

It's also usual to take in the full path.

tantona commented:
@brian-brazil thank you again for your reviews!

One more for you - hopefully I've covered all the bases. I'm also in the hangops prometheus slack channel

brian-brazil commented on generator/README.md:
```diff
@@ -24,6 +24,19 @@ The generator reads in from `generator.yml` and writes to `snmp.yml`.
+docker run -ti \
+  -v $PWD/mibs:/root/.snmp/mibs \
+  -v $PWD/generator.yml:/opt/generator.yml:ro \
+  -v $PWD/out/:/opt/build \
```
Why put it in a subdir of the cwd rather than directly in the cwd?

tantona replied:
good call - making this change

brian-brazil commented on generator/README.md:
````diff
@@ -24,6 +24,19 @@ The generator reads in from `generator.yml` and writes to `snmp.yml`.
+```
+docker build -t snmp-generator .
+docker run -ti \
+  -v $PWD/mibs:/root/.snmp/mibs \
````
Usually mibs would be in $HOME/.snmp/mibs

tantona replied:
/root is $HOME in this container

brian-brazil replied:
I meant on the LHS

tantona replied:
Oh I see - For my use case I'm pulling down vendor specific mibs for network devices - so there is no canonical mib location

I'm not opposed to changing this to $HOME/.snmp/mibs if that works better for you

brian-brazil replied:
That's where net-snmp will usually be looking, so where I'd expect users to have them.

tantona replied:
done

brian-brazil commented on generator/main.go:
```diff
@@ -46,19 +63,20 @@ func generateConfig(nodes *Node, nameToNode map[string]*Node) {
 
 var (
 	generateCommand    = kingpin.Command("generate", "Generate snmp.yml from generator.yml")
+	outputDir          = generateCommand.Flag("output-dir", "Directory to write resulting config file").Short('o').String()
```
You should take a path, not a directory

tantona commented:
@brian-brazil back at ya

brian-brazil commented on generator/README.md:
```diff
@@ -31,7 +31,7 @@ If you would like to run the generator in docker to generate your `snmp.yml` con
 docker build -t snmp-generator .
 docker run -ti \
-  -v $PWD/mibs:/root/.snmp/mibs \
+  -v $$HOME/.snmp/mibs:/root/.snmp/mibs \
```
You only need one $

brian-brazil commented on generator/main.go:
```diff
@@ -46,19 +63,20 @@ func generateConfig(nodes *Node, nameToNode map[string]*Node) {
 
 var (
 	generateCommand    = kingpin.Command("generate", "Generate snmp.yml from generator.yml")
+	outputPath         = generateCommand.Flag("output-path", "Path to to write resulting config file").Short('o').String()
```
The default should be snmp.yml

brian-brazil commented:
Thanks!
